bpf/lib/encrypt.h

```diff
@@ -204,6 +205,120 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ENDPOINT_ROUTES */
 }
 
+/* checks whether a IPsec redirect should be performed for the security id
+ * we do not IPsec encrypt:
+ * 1. Host-to-Host or Pod-to-Host traffic
+ * 2. Traffic leaving the cluster
+ * 3. Remote nodes including Kube API server
+ */
+static __always_inline int
+ipsec_redirect_sec_id_ok(__u32 sec_id) {
+	if (sec_id == HOST_ID)
+		return 0;
+	if (!identity_is_cluster(sec_id))
+		return 0;
+	if (identity_is_remote_node(sec_id))
+		return 0;
+	return 1;
+}
+
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info __maybe_unused *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	struct ipv6hdr __maybe_unused *ip6;
+	__u32 magic __maybe_unused = 0;
+	int ret = 0;
+	union macaddr dst_mac = CILIUM_NET_MAC;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+
+#  if defined(TUNNEL_MODE)
+		/* tunnel mode needs a bit of special handling when
+		 * encapsulated packets get here the destination address is
+		 * already a cluster node IP.
+		 *
+		 * cluster node IPs will have 0.0.0.0 as their tunnel_endpoint
+		 * fields in the ipcache. However, IPsec needs the NodeID of the
+		 * destination from the NodeMap. We can pass in the destination
+		 * of this VXLAN packet as the tunnel_endpoint field to
+		 * accomplish this.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), false,
+						true);
+			if (ret != CTX_ACT_OK)
+				return ret;
+			goto overlay_encrypt;
+		}
+#  endif /* TUNNEL_MODE */
+
+		break;
+# endif /* ENABLE_IPV4 */
+
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+#ifndef TUNNEL_MODE
+		/* handle native routing ipv6 */
+		if (!revalidate_data(ctx, &data, &data_end, &ip6))
+			return DROP_INVALID;
+
+		dst = lookup_ip6_remote_endpoint((union v6addr *)&ip6->daddr, 0);
+		src = lookup_ip6_remote_endpoint((union v6addr *)&ip6->saddr, 0);
+
+		break;
+#endif /* TUNNEL_MODE */
+# endif /* ENABLE_IPv6 */
+	default:
+		return CTX_ACT_OK;
+	}
+
+	if (!dst || !src)
+		return CTX_ACT_OK;
+
+	if (!ipsec_redirect_sec_id_ok(dst->sec_identity) ||
+	    !ipsec_redirect_sec_id_ok(src->sec_identity))
+		return CTX_ACT_OK;
+
+	/* mark packet for encryption based on ipcache's key */
+	ret = set_ipsec_encrypt(ctx, dst->key, dst->tunnel_endpoint,
+				src->sec_identity, false, true);
```

Comment:
`set_ipsec_encrypt(....., use_meta=false, use_spi_from_map=true)`:

when it's called with `use_spi_from_map=true`, the 2nd argument `spi=dst->key` is not used at all, right? we'll always get spi from node_map.

though this is https://github.com/cilium/cilium/issues/34790, I'm still concerned this step is too big: moving ipsec hooks to eth0 and using spi from node_map instead of ipcache in the meantime seems difficult in terms of risk control.

Reply:
@jschwinger233 so you would rather move to using node-map more, later? Not during this change?

Reply:
> you would rather move to using node-map more, later? Not during this change?

Yeah. I'm not opposing it, just trying to make this change noticed and reviewed by more engineers, increasing the chances of catching potential bugs.

Reply:
Let's use node map in a separate PR shall we? My points are:
1. Neither PR description or commit message mentioned this change
2. We expect only tunnel routing is affected by this PR
3. It's "hidden" in a big commit that engineers are easy to overlook
4. This looks like the last place to use ipcache encryptkey field for ipsec, we can also delete the obsolete control plane code once switching to node map

Reply:
@jschwinger233 Yup, lets move to node map then. I'll work on this commit after I finish up getting the v1.17 backport branch in shape.

Reply:
Running tests for this change now, including backport branch.